src/hotspot/share/opto/vectorIntrinsics.cpp

```diff
@@ -376,6 +376,227 @@ bool LibraryCallKit::inline_vector_nary_operation(int n) {
+    }
+    return false; // not supported
+  }
+
+  Node* opd1 = NULL; Node* opd2 = NULL; Node* opd3 = NULL;
+  switch (n) {
+    case 3: {
+      opd3 = unbox_vector(argument(7), vbox_type, elem_bt, num_elem);
+      if (opd3 == NULL) {
+        if (C->print_intrinsics()) {
+          tty->print_cr("  ** unbox failed v3=%s",
+                        NodeClassNames[argument(7)->Opcode()]);
+        }
+        return false;
+      }
+      // fall-through
+    }
+    case 2: {
+      opd2 = unbox_vector(argument(6), vbox_type, elem_bt, num_elem);
+      if (opd2 == NULL) {
+        if (C->print_intrinsics()) {
+          tty->print_cr("  ** unbox failed v2=%s",
+                        NodeClassNames[argument(6)->Opcode()]);
+        }
+        return false;
+      }
+      // fall-through
+    }
+    case 1: {
+      opd1 = unbox_vector(argument(5), vbox_type, elem_bt, num_elem);
+      if (opd1 == NULL) {
+        if (C->print_intrinsics()) {
+          tty->print_cr("  ** unbox failed v1=%s",
+                        NodeClassNames[argument(5)->Opcode()]);
+        }
+        return false;
+      }
+      break;
+    }
+    default: fatal("unsupported arity: %d", n);
+  }
+
+  Node* mask = NULL;
+  if (use_mask) {
+    ciKlass* mbox_klass = mask_klass->const_oop()->as_instance()->java_lang_Class_klass();
+    assert(is_vector_mask(mbox_klass), "argument(2) should be a mask class");
+    const TypeInstPtr* mbox_type = TypeInstPtr::make_exact(TypePtr::NotNull, mbox_klass);
+    mask = unbox_vector(argument(n + 5), mbox_type, elem_bt, num_elem);
+    if (mask == NULL) {
+        if (C->print_intrinsics()) {
+          tty->print_cr("  ** unbox failed mask=%s",
+                        NodeClassNames[argument(n + 5)->Opcode()]);
+        }
+        return false;
+      }
+  }
+
+  Node* operation = NULL;
+  if (opc == Op_CallLeafVector) {
+    assert(UseVectorStubs, "sanity");
+    operation = gen_call_to_svml(opr->get_con(), elem_bt, num_elem, opd1, opd2);
+    if (operation == NULL) {
+      if (C->print_intrinsics()) {
+        tty->print_cr("  ** svml call failed");
+      }
+      return false;
+     }
+  } else {
+    const TypeVect* vt = TypeVect::make(elem_bt, num_elem);
+    switch (n) {
+      case 1:
+      case 2: {
+        operation = VectorNode::make(sopc, opd1, opd2, vt);
+        break;
+      }
+      case 3: {
+        operation = VectorNode::make(sopc, opd1, opd2, opd3, vt);
+        break;
+      }
+      default: fatal("unsupported arity: %d", n);
+    }
+  }
+
+  // Currently just focus on binary mask operation.
+  if (n == 2 && mask != NULL) {
+    if (sopc != 0 &&
+        Matcher::match_rule_supported_masked_vector(sopc, num_elem, elem_bt) &&
+        Matcher::match_rule_supported(Op_VectorToMask)) {
+      const TypeVMask* vmask_type = TypeVMask::make(elem_bt, num_elem);
+      mask = gvn().transform(new VectorToMaskNode(mask, vmask_type));
+      operation->add_req(mask);
```

Comment:
We are adding a new input to a binary operation with existing opcodes i.e.
Eg. following will be graph shapes with or w/o mask node.
   AddVI SRC1 SRC2
   AddVI SRC1 SRC2 MASK

We may be able to curtail the number of instruction patterns for masked operation by folding blend + vector operation into a new VectorMaskedOperNode.  Effectively this way we may be able to remove lots of redundant instruction patterns from the AD files, which has been a cause of concern for us(X86) in past.

Reply:
One downside of `AddVI SRC1 SRC2 MASK` shape is that the operation is not commutative anymore:
```
AddVI SRC1 SRC2 MASK = VectorBlend SRC1 (AddVI SRC1 SRC2) MASK
AddVI SRC2 SRC1 MASK = VectorBlend SRC2 (AddVI SRC2 SRC1) MASK
```

Reply:
Thanks for looking at this PT @iwanowww ! Yes, I agree that this can change the commutative feature by appending the mask. I found a potential influence during matching. The codes look like:
```
  void MatchNode::count_commutative_op(int& count) {
  static const char *commut_op_list[] = {
    "AddI","AddL","AddF","AddD",
    "AddVB","AddVS","AddVI","AddVL","AddVF","AddVD",
    "AndI","AndL",
    "AndV",
    "MaxI","MinI","MaxF","MinF","MaxD","MinD",
    "MaxV", "MinV",
    "MulI","MulL","MulF","MulD",
    "MulVB","MulVS","MulVI","MulVL","MulVF","MulVD",
    "MinV","MaxV",
    "OrI","OrL",
    "OrV",
    "XorI","XorL",
    "XorV"
  };
  int cnt = sizeof(commut_op_list)/sizeof(char*);

  if( _lChild && _rChild && (_lChild->_lChild || _rChild->_lChild) ) {
    // Don't swap if right operand is an immediate constant.
    bool is_const = false;
    if( _rChild->_lChild == NULL && _rChild->_rChild == NULL ) {
      FormDict &globals = _AD.globalNames();
      const Form *form = globals[_rChild->_opType];
      if ( form ) {
        OperandForm  *oper = form->is_operand();
        if( oper && oper->interface_type(globals) == Form::constant_interface )
          is_const = true;
      }
    }
    if( !is_const ) {
      for( int i=0; i<cnt; i++ ) {
        if( strcmp(_opType, commut_op_list[i]) == 0 ) {
          count++;
          _commutative_id = count; // id should be > 0
          break;
        }
      }
    }
  }
  if( _lChild )
    _lChild->count_commutative_op(count);
  if( _rChild )
    _rChild->count_commutative_op(count);
 }
```
For such issues, is it possible to add the `commutative` feature to a matchnode, or adding more conditions when checking whether the node is commutative? Regarding to the masked vector representation, do you have any better idea? We also tried with the `VectorMaskWrapperNode` before, but it's not better with some reasons that I pointed out at page-5 in http://cr.openjdk.java.net/~xgong/rfr/mask/Vector%20API%20masking%20support%20proposal%20for%20Arm%20SVE.pdf.  Thanks!